finat/tensor_product.py

```diff
@@ -0,0 +1,177 @@
+        # tensor:
+        #
+        #   \grad \grad \grad u
+        #
+        # A multiindex can denote an entry of this derivative tensor.
+        # For example, i = (1, 0, 1) refers deriving `u` first in the
+        # y-direction (second direction, indexing starts at zero),
+        # then in the x-direction, and finally in the second direction
+        # again.  Usually the order of these derivatives does not
+        # matter, so the derivative tensor is symmetric.  For specific
+        # conditions about when this actually holds, please refer to
+        # your favourite calculus textbook.  We call `i` here a
+        # "derivative tensor multiindex", or tensor multiindex for
+        # short.  For example, the following tensor multiindices
+        # correspond to the (1, 2) canonical multiindex:
+        #
+        #   (0, 1, 1)
+        #   (1, 0, 1)
+        #   (1, 1, 0)
+        #
+        # Now you should be ready to get go.
+
+        # Spatial dimension
+        dimension = self.cell.get_spatial_dimension()
+        # The derivative tensor
+        tensor = numpy.empty((dimension,) * derivative, dtype=object)
+        # The identity matrix, used to facilitate conversion from
+        # tensor multiindex to canonical multiindex form.
+        eye = numpy.eye(dimension, dtype=int)
+        # A list of multiindices, one multiindex per subelement, each
+        # multiindex describing the shape of basis functions of the
+        # subelement.
+        alphas = [fe.get_indices() for fe in self.factors]
+        # A list of slices that are used to select dimensions
+        # corresponding to each subelement.
+        dim_slices = TensorProductCell._split_slices([c.get_spatial_dimension()
+                                                      for c in self.cell.cells])
+        # 'delta' is a tensor multiindex consisting of only fixed
+        # indices for populating the entries of the derivative tensor.
+        for delta in numpy.ndindex(tensor.shape):
+            # Get the canonical multiindex corresponding to 'delta'.
+            D_ = tuple(eye[delta, :].sum(axis=0))
+            # Split this canonical multiindex for the subelements.
+            Ds = [D_[s] for s in dim_slices]
+            # GEM scalars (can have free indices) for collecting the
+            # contributions from the subelements.
+            scalars = []
+            for fe, ps_, e, D, alpha in zip(self.factors, ps_factors, entities, Ds, alphas):
+                # Ask the subelement to tabulate at the required derivative order.
+                value = fe.basis_evaluation(ps_, entity=e, derivative=sum(D))
+                # Nice, but now we have got a subelement derivative
+                # tensor of the given order, while we only need a
+                # specific derivative.  So we convert the subelement
+                # canonical multiindex to tensor multiindex.
+                d = tuple(chain(*[(dim,) * count for dim, count in enumerate(D)]))
```

Comment:
So you have to do this dance between tensor and canonical multi index because it's easier to decide which derivatives apply to which sub element in the tensor case, but everything else wants the canonical case, right?

I think like the reader of this code's brain could be eased if you wrapped some of these iterable one-liners up in functions that say what they do: e.g. `d = to_tensor(D)` and earlier `D_ = to_canonical(delta)`...